src/librustc_codegen_ssa/back/write.rs

```diff
@@ -538,6 +560,32 @@ fn produce_final_output_artifacts(sess: &Session,
                 // (We don't have to do any work in this case.)
             }
         }
+        if sess.opts.debugging_opts.no_link {
+            if let Some(ref module) = compiled_modules.allocator_module {
+                let module_name = Some(&module.name[..]);
+                let path = crate_output.temp_path(output_type, module_name);
+                let extension = format!("allocator.{}", output_type.extension());
+
+                copy_gracefully(&path,
+                                &crate_output.path(output_type).with_extension(extension));
+                if !sess.opts.cg.save_temps && !keep_numbered {
+                    // The user just wants `foo.x`, not `foo.#module-name#.x`.
+                    remove(sess, &path);
+                }
+            }
+            if let Some(ref module) = compiled_modules.metadata_module {
```

Comment:
The structure of this code looks similar to the bit above (and same below); it would be good to avoid copy paste programming.

Reply:
Totally agreed, but not doing so would require refactoring existing code a bit. They are actually copied from existing code.
I'd like to avoid refactoring and new feature in the same commit or pull request. Let me create a separate PR for the refactoring and then rebase this PR.

Reply:
Certainly. I only care that the work to refactor eventually gets done. Feel free to structure your refactorings as you see fit over PRs / commits. :)

Reply:
I've updated the changes based on reviewers' feedbacks, and this file won't be touched anymore.